### ISSUE TITLE
Proposal: Allow containers to "steal" already allocated host port

### ISSUE THREAD

**Issue by tomfotherby** [docker/docker#14945]
Currently (i.e. *docker v1.7.1*), if you try to start a container that uses the same host port as a already running container you get a error:  :no_entry: **failed: port is already allocated**. Example:

````
$docker run -p 8080:80 -d nginx
c851f7955d48efae7455649922285c185560e39d44de0b618dd31e3437cb4623

$docker run -p 8080:80 -d nginx
b6e8728e24c56bcdf24a69fae0c95de7d31f1c4fa6216b257dab10f175699a93
Error response from daemon: Cannot start container b6e8728e24c56bcdf24a69fae0c95de7d31f1c4fa6216b257dab10f175699a93: Bind for 0.0.0.0:8080 failed: port is already allocated
```

It would be very useful if there was a option to "force" the start of the 2nd container. Maybe the original container could be stopped or paused or be given another free port (like if `-P` was used).

Why would this be useful? *i.e. why can't you just stop the other container first?*
If you are running a webapp in a container and getting many requests per second and want to do a upgrade because you have a new version of the containerised webapp, then stopping the container and starting the new version will result in a little downtime.
This "port stealing" feature would allow zero-downtime deployments without needing a proxy (or whatever you want to call it, e.g. load-balancer , or [ambassador](http://docs.docker.com/articles/ambassador_pattern_linking/) , or [jumper](https://coreos.com/blog/jumpers-and-the-software-defined-localhost/), etc. hipache, [nginx-proxy](https://github.com/jwilder/nginx-proxy), etc).

**Comment by thaJeztah** [comment]
I'm -1 on this. I actually think this will lead to a *longer* downtime than using the existing solutions (e.g. Hipache or [nginx-proxy](https://github.com/jwilder/nginx-proxy)). What if the new container needs some time before it's ready to accept connections, and access to the old container is already removed (I.e. The port is unmapped). What should happen if the new container *fails* to start?

The existing solutions cover this by allowing *multiple* containers handling the same domain/website/app. After starting the new container, you can safely remove the old container without downtime.

**Comment by cpuguy83** [comment]
Sorry, I agree with @thaJeztah.
-1
This is why load balancers exist.
If it's critical enough to need 0 downtime, then it ought to be setup as such.

**Comment by tomfotherby** [comment]
Ok, thanks for the consideration, I appreciate the comments. I'll close the issue based on the negative consensus.

**Comment by tomfotherby** [comment]
I wonder whether it'll be possible (in the future) for me to create a docker plugin that might be able to do this to shield everyone from the negative affects but enable it for folks that want to try it. I'll keep a eye out for tutorials and documentation about plugins...

**Comment by thaJeztah** [comment]
There's already some documentation for network plugins in the "experimental" documentation, but it's still WIP, so specifications may change: https://github.com/docker/docker/tree/master/experimental


Wondering what the difficulties were that you ran into. Using a simple front-end proxy isn't difficult.

I like using nginx-proxy; although it may not be the most advanced solution, it's really simple to use.

All you need to do is start the proxy:

```Bash
docker run \
        -d \
        -p 80:80 \
        -p 443:443 \
        --restart always \
        -t --name nginx_proxy \
        -v /var/run/docker.sock:/tmp/docker.sock:ro \
        jwilder/nginx-proxy:latest
```

And add a `VIRTUAL_HOST` environment-variable to each website you want to make available through the proxy. The proxy will route traffic to that container, based on the host-name(s) you provided and automatically rotate traffic if multiple containers have specified the same host-name(s). You don't even have to "publish" the ports (they only have to be "exposed" / `EXPOSE`);

```Bash
docker run -d -e VIRTUAL_HOST=app.example.com,www.app.example.com --name=myapp-v1.0.0 nginx
```

To upgrade your app, start a container with the new version and stop (and/or remove) the old container;

```Bash
docker run -d -e VIRTUAL_HOST=app.example.com,www.app.example.com --name=myapp-v1.0.1 nginx
docker stop myapp-previous
```

In case you discovered the updated app is bad, and want to revert, simply;

```Bash
docker start myapp-v1.0.0
docker stop myapp-v1.0.1
docker rm myapp-v1.0.1
```

**Comment by tomfotherby** [comment]
Thanks @thaJeztah , 

> Wondering what the difficulties were that you ran into?

It's not that we ran into any major difficulties. It's just that the front-end proxy is a extra moving part and if docker allowed me to always steal port 80 for my new containers then it would make life that little bit simpler and more magical :rainbow: .

Since you ask, these are the difficulties we ran into using a front-end proxy, nothing insolvable:

 1. Rather than cyan releases like you describe with [nginx-proxy](https://github.com/jwilder/nginx-proxy) above, we have a preference for blue-green deployments with a clean cutover to the new app.
 1. We run crons within the currently active app using `docker exec`. To figure out which container is currently servicing requests, we have some small scripts to figure it out - it would be a little easier to just always do `docker exec` to the container at port 80.
 1. We found our proxy when working in a Vagrant environment wan't working ([Source](https://docs.vagrantup.com/v2/synced-folders/virtualbox.html)). Turned out we needed to set `sendfile off` on nginx. But this isn't advised for production.
